File: frontend/src/lib/api/deepsearch.ts
/**
 * DeepSearch RAG API Client
 * Handles chatbot-style queries to the RAG Sentinel agent.
 */

import { apiClient } from "./client";

export interface DeepSearchRequest {
  query: string;
  top_k?: number;
}

export interface Citation {
  filename: string;
  page: number | null;
  confidence: number | null;
}

export interface SourceSummary {
  filename: string;
  pages: number[];
  confidence: number | null;
}

export interface DeepSearchResponse {
  query: string;
  answer: string;
  citations: Citation[];
  source_summary: SourceSummary[];
  error: string | null;
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  timestamp: Date;
  citations?: Citation[];
  sources?: SourceSummary[];
  isLoading?: boolean;
  error?: string;
}

/**
 * Perform DeepSearch RAG query
 * Uses the File Vault documents as the knowledge base
 */
export async function deepSearch(
  query: string,
  topK: number = 5
): Promise<DeepSearchResponse> {
  if (!query || query.trim().length < 3) {
    throw new Error("Query must be at least 3 characters long");
  }
  if (topK < 1 || topK > 50) {
    throw new Error("top_k must be between 1 and 50");
  }

  const response = await apiClient.post<DeepSearchResponse>(
    "/api/v1/deepsearch/",
    {
      query: query.trim(),
      top_k: topK,
    }
  );
  return response.data;
}

/**
 * Generate a unique message ID
 */
export function generateMessageId(): string {
  return `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

File: backend/app/api/v1/endpoints/deepsearch.py
"""
DeepSearch-RAG-Sentinel API Endpoints
Hyper-accurate, cited answers using OpenRouter/DeepSeek API.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from app.core.security import User, get_current_active_user
from app.services.deepsearch.sentinel_agent import DeepSearchRAGSentinel

router = APIRouter()


# Request/Response Models
class DeepSearchRequest(BaseModel):
    """DeepSearch query request model."""
    query: str
    top_k: int = 5


class Citation(BaseModel):
    """Citation model."""
    filename: str
    page: int | None = None
    confidence: float | None = None


class SourceSummary(BaseModel):
    """Source summary model."""
    filename: str
    pages: list[int] = []
    confidence: float | None = None


class DeepSearchResponse(BaseModel):
    """DeepSearch response model."""
    query: str
    answer: str
    citations: list[Citation] = []
    source_summary: list[SourceSummary] = []
    error: str | None = None


@router.post("/", response_model=DeepSearchResponse, status_code=status.HTTP_200_OK)
async def deepsearch_query(
    request: DeepSearchRequest,
    current_user: User = Depends(get_current_active_user),
) -> DeepSearchResponse:
    """
    Perform DeepSearch-RAG-Sentinel query using OpenRouter/DeepSeek API.
    
    This endpoint provides hyper-accurate, cited answers by:
        - Using OpenRouter/DeepSeek API for document retrieval
    - Applying Layout-Aware Parsing for tables, charts, and diagrams
    - Including citations in format [Source: Filename, Page X]
    - Never speculating if information is not in the PDFs
    
    Args:
        request: DeepSearch request with query
        current_user: Current authenticated user
        
    Returns:
        DeepSearch response with answer and citations
        
    Raises:
        HTTPException: If query fails
    """
    # Validate query
    if not request.query or len(request.query.strip()) < 3:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Query must be at least 3 characters long"
        )
    
    # Validate top_k
    if request.top_k < 1 or request.top_k > 50:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="top_k must be between 1 and 50"
        )
    
    # Initialize DeepSearch-RAG-Sentinel agent
    agent = DeepSearchRAGSentinel()
    
    try:
        # Perform DeepSearch query
        result = await agent.query(
            user_query=request.query,
            top_k=request.top_k,
        )
        
        # Format citations
        citations = [
            Citation(
                filename=c.get("filename", "Unknown"),
                page=c.get("page"),
                confidence=c.get("confidence"),
            )
            for c in result.get("citations", [])
        ]
        
        # Format source summary
        source_summary = [
            SourceSummary(
                filename=s.get("filename", "Unknown"),
                pages=s.get("pages", []),
                confidence=s.get("confidence"),
            )
            for s in result.get("source_summary", [])
        ]
        
        return DeepSearchResponse(
            query=result["query"],
            answer=result["answer"],
            citations=citations,
            source_summary=source_summary,
            error=result.get("error"),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"DeepSearch query failed: {str(e)}"
        ) from e

